build-stack.sh

```diff
@@ -3,45 +3,40 @@
 # Script for building Kythe index of Haskell packages.
 # Usage: bash build-stack.sh /tmp/logs package...
 
-fail() {
-  (>&2 echo "$1")
+if (($# < 2)); then
+  echo "Usage: $0 /tmp/logs package..." >&2
   exit 1
-}
+fi
+
+if [[ -z "$STACK_YAML" ]]; then
+  echo 'Please set $STACK_YAML' >&2
+  exit 1
+fi
 
 # Directory where to build the index.
 export INDEXER_OUTPUT_DIR=$1
+mkdir -p "$INDEXER_OUTPUT_DIR"
 
-# REALGHC is used by stack wrapper ghc. Note that it must be set before
-# altering the PATH.
-export REALGHC=$(stack path --compiler-exe)
+project_root=$(cd "$(dirname "$0")"; pwd)
 
 # Build and index the packages
 # ============================
-# Put stack wrapper ghc script, ghc-pkg (from compiler-bin) and
-# ghc_kythe_wrapper (from local-install-root) on the PATH. Note that stack
-# wrapper ghc script replaces the system ghc.
-PATH=$PWD/wrappers/stack:$(stack path --compiler-bin):$PATH:$(stack path --local-install-root)/bin \
-  stack --system-ghc build --no-nix --force-dirty ${@:2}
-[[ $? != 0 ]] && fail "Indexing failed!"
-
-# Serve the index
-# ===============
-# It's probably more efficient to cat them together, but this way we see
-# if a given one is corrupted for any reason.
-for e in $(ls ${INDEXER_OUTPUT_DIR}/*entries)
-do
-  echo " * ${e}"
-  cat ${e} | /opt/kythe/tools/write_entries --graphstore ${INDEXER_OUTPUT_DIR}/gs
+# Add a post hook to stack ghc to invoke our post_hook which in turn invokes ghc_kythe_wrapper.
+# Since there seems no easy way to hook the compiler used by stack,
+# we modify $compiler_exe in place and recover it on exit from the shell script.
+compiler_exe=$(stack path --compiler-exe)
+if ! grep -q wrapper "$compiler_exe"; then
+  cp -f "$compiler_exe" "$compiler_exe.orig"
+  sed -i 's,^exec .*,"$executablename" -B"$topdir" "$@" \&\& '"$project_root/wrappers/stack/ghc_post_hook"' "$@",' "$compiler_exe"
```

Comment:
Changed the comment:
```bash
# Build and index the packages
# ============================
# Add a post hook to $compiler_exe to invoke our post_hook which in turn invokes ghc_kythe_wrapper.
# Since there seems no easy way to hook the compiler used by stack,
# we modify $compiler_exe (something like ~/.stack/programs/x86_64-linux/ghc-8.0.2/bin/ghc)
# in place and recover it on exit from the shell script.
compiler_exe=$(stack path --compiler-exe)
if ! grep -q wrapper "$compiler_exe"; then
  cp -f "$compiler_exe" "$compiler_exe.orig"
  # Replace the `exec` line with `exec ghc -B... "$@" && ghc_post_hook "$@"`
  sed -i 's,^exec .*,"$executablename" -B"$topdir" "$@" \&\& '"$project_root/wrappers/stack/ghc_post_hook"' "$@",' "$compiler_exe"
  trap "mv -f $compiler_exe.orig $compiler_exe" EXIT
fi
```